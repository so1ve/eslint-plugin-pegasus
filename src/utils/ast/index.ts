export * from "./call-or-new-expression";
export * from "./is-member-expression";
export * from "./is-method-call";
