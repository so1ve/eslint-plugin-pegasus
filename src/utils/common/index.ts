export * from "./boolean";
export * from "./is-function-self-used-inside";
export * from "./is-logical-expression";
export * from "./is-node-matches";
export * from "./is-node-value-not-function";
export * from "./is-number";
export * from "./is-same-identifier";
export * from "./parentheses";
export * from "./rule";
