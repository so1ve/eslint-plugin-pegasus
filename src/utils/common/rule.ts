import type { TSESLint } from "@typescript-eslint/utils";
import type { RuleCreateFunction } from "@typescript-eslint/utils/ts-eslint";

const isIterable = (object: any): object is Iterable<any> =>
	typeof object?.[Symbol.iterator] === "function";

class FixAbortError extends Error {}
const fixOptions = {
	abort() {
		throw new FixAbortError("Fix aborted.");
	},
};

export type AbortFunction = () => void;

type WrappedFixFunction = (
	fixer: TSESLint.RuleFixer,
	additional: { abort: AbortFunction },
) =>
	| IterableIterator<TSESLint.RuleFix>
	| TSESLint.RuleFix
	| readonly TSESLint.RuleFix[]
	| null;

export const wrapFixFunction =
	(fix: WrappedFixFunction): TSESLint.ReportFixFunction =>
	(fixer) => {
		const result = fix(fixer, fixOptions);

		if (isIterable(result)) {
			try {
				return [...result];
			} catch (error) {
				if (error instanceof FixAbortError) {
					return [];
				}

				throw error;
			}
		}

		return result;
	};

type Listener = (node: any) => void;
type EnhanceContext<CreateFunction extends RuleCreateFunction> =
	Parameters<CreateFunction>[0] & {
		on: (selectorOrSelectors: string | string[], listener: Listener) => void;
		onExit: (
			selectorOrSelectors: string | string[],
			listener: Listener,
		) => void;
	};
export type EnhancedAnyContext = EnhanceContext<RuleCreateFunction>;
type EnhanceCreateFunction<CreateFunction extends RuleCreateFunction> = (
	context: EnhanceContext<CreateFunction>,
) => ReturnType<CreateFunction> | void;
export function wrapCreateFunction<
	Options extends readonly unknown[],
	MessageIds extends string,
	CreateFunction extends RuleCreateFunction<MessageIds>,
>(
	create: EnhanceCreateFunction<CreateFunction>,
): RuleCreateFunction<MessageIds, Options> {
	const wrapped: RuleCreateFunction<MessageIds, Options> = (context) => {
		const listeners: Record<string, (Listener | undefined)[]> = {};
		function addListener(selector: string, listener: Listener | undefined) {
			listeners[selector] ??= [];
			listeners[selector].push(listener);
		}

		const contextProxy: EnhanceContext<CreateFunction> = new Proxy(context, {
			get(target, property, receiver) {
				if (property === "on") {
					return (
						selectorOrSelectors: string | string[],
						listener: Listener,
					) => {
						const selectors = Array.isArray(selectorOrSelectors)
							? selectorOrSelectors
							: [selectorOrSelectors];
						for (const selector of selectors) {
							addListener(selector, listener);
						}
					};
				}

				if (property === "onExit") {
					return (
						selectorOrSelectors: string | string[],
						listener: Listener,
					) => {
						const selectors = Array.isArray(selectorOrSelectors)
							? selectorOrSelectors
							: [selectorOrSelectors];
						for (const selector of selectors) {
							addListener(`${selector}:exit`, listener);
						}
					};
				}

				return Reflect.get(target, property, receiver);
			},
		}) as any;

		for (const [selector, listener] of Object.entries(
			create(contextProxy) ?? {},
		)) {
			addListener(selector, listener as any);
		}

		return Object.fromEntries(
			Object.entries(listeners).map(([selector, listeners]) => [
				selector,
				// Listener arguments can be `codePath, node` or `node`
				(...listenerArguments) => {
					for (const listener of listeners) {
						listener?.(...listenerArguments);
					}
				},
			]),
		);
	};

	return wrapped;
}
