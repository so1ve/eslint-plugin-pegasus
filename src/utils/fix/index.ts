export * from "./remove-member-expression-property";
export * from "./remove-method-call";
