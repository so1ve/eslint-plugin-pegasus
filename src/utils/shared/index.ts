export * from "./simple-array-search-rule";
