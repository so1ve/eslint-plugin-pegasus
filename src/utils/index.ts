import { ESLintUtils } from "@typescript-eslint/utils";

export const createRule = ESLintUtils.RuleCreator(
	(ruleName) =>
		`https://github.com/so1ve/eslint-plugin-pegasus/blob/main/docs/rules/${ruleName}.md`,
);
