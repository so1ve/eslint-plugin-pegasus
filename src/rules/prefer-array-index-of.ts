import { createRule } from "../utils";
import { wrapCreateFunction } from "../utils/common";
import { simpleArraySearchRule } from "../utils/shared";

export const RULE_NAME = "prefer-array-index-of";
export type MessageIds = string;
export type Options = [];

const indexOfOverFindIndexRule = simpleArraySearchRule({
	method: "findIndex",
	replacement: "indexOf",
});
const lastIndexOfOverFindLastIndexRule = simpleArraySearchRule({
	method: "findLastIndex",
	replacement: "lastIndexOf",
});

export default createRule<Options, MessageIds>({
	name: RULE_NAME,
	meta: {
		type: "suggestion",
		docs: {
			description:
				"Prefer `Array#{indexOf,lastIndexOf}()` over `Array#{findIndex,findLastIndex}()` when looking for the index of an item.",
			requiresTypeChecking: true,
		},
		fixable: "code",
		schema: [],
		messages: {
			...indexOfOverFindIndexRule.messages,
			...lastIndexOfOverFindLastIndexRule.messages,
		},
		hasSuggestions: true,
	},
	defaultOptions: [],
	create: wrapCreateFunction((context) => {
		indexOfOverFindIndexRule.listen(context);
		lastIndexOfOverFindLastIndexRule.listen(context);
	}),
});
