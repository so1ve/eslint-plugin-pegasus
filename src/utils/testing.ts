import { resolve } from "node:path";

import { RuleTester } from "../../vendor/rule-tester/src/RuleTester";

const fixturesDir = resolve(process.cwd(), "test", "fixtures");

export function getRuleTester() {
	const ruleTester = new RuleTester({
		parser: require.resolve("@typescript-eslint/parser"),
		parserOptions: {
			ecmaVersion: 2018,
			tsconfigRootDir: fixturesDir,
			project: "./tsconfig.json",
		},
	});

	return ruleTester;
}
