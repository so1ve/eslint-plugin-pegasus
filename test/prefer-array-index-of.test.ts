import rule, { RULE_NAME } from "../src/rules/prefer-array-index-of";
import { testRuleSnapshot } from "../src/utils/testing";
import { createSimpleArraySearchRuleTestFixtures } from "./fixtures/prefer-array-index-of";

const indexOfOverFindIndexFixtures = createSimpleArraySearchRuleTestFixtures({
	method: "findIndex",
	replacement: "indexOf",
});

testRuleSnapshot(RULE_NAME, rule, indexOfOverFindIndexFixtures);

const lastIndexOfOverFindLastIndexFixtures =
	createSimpleArraySearchRuleTestFixtures({
		method: "findLastIndex",
		replacement: "lastIndexOf",
	});

testRuleSnapshot(RULE_NAME, rule, lastIndexOfOverFindLastIndexFixtures);
