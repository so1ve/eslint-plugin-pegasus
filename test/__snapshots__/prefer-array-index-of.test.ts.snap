// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefer-array-index-of > invalid > // 1
(0, values)
\\u0009// 2
\\u0009./* 3 */findIndex /* 3 */ (
\\u0009\\u0009/* 4 */
\\u0009\\u0009x /* 5 */ => /* 6 */ x /* 7 */ === /* 8 */ "foo" /* 9 */
\\u0009) /* 10 */ 1`] = `
"// 1
(0, values)
	// 2
	./* 3 */indexOf /* 3 */ (
		/* 4 */
		\\"foo\\" /* 9 */
	) /* 10 */"
`;

exports[`prefer-array-index-of > invalid > // 1
(0, values)
\\u0009// 2
\\u0009./* 3 */findLastIndex /* 3 */ (
\\u0009\\u0009/* 4 */
\\u0009\\u0009x /* 5 */ => /* 6 */ x /* 7 */ === /* 8 */ "foo" /* 9 */
\\u0009) /* 10 */ 1`] = `
"// 1
(0, values)
	// 2
	./* 3 */lastIndexOf /* 3 */ (
		/* 4 */
		\\"foo\\" /* 9 */
	) /* 10 */"
`;

exports[`prefer-array-index-of > invalid > foo.findIndex(function (element) {
\\u0009return element === bar.findIndex(x => x === 1);
}); 1`] = `
"foo.findIndex(function (element) {
	return element === bar.indexOf(1);
});"
`;

exports[`prefer-array-index-of > invalid > foo.findIndex(function a(x) {
\\u0009return x === (function (a) {
\\u0009\\u0009return a(this) === arguments[1]
\\u0009}).call(thisObject, anotherFunctionNamedA, secondArgument)
}) 1`] = `
"foo.findIndex(function a(x) {
	return x === (function (a) {
		return a(this) === arguments[1]
	}).call(thisObject, anotherFunctionNamedA, secondArgument)
})"
`;

exports[`prefer-array-index-of > invalid > foo.findLastIndex(function (element) {
\\u0009return element === bar.findLastIndex(x => x === 1);
}); 1`] = `
"foo.findLastIndex(function (element) {
	return element === bar.lastIndexOf(1);
});"
`;

exports[`prefer-array-index-of > invalid > foo.findLastIndex(function a(x) {
\\u0009return x === (function (a) {
\\u0009\\u0009return a(this) === arguments[1]
\\u0009}).call(thisObject, anotherFunctionNamedA, secondArgument)
}) 1`] = `
"foo.findLastIndex(function a(x) {
	return x === (function (a) {
		return a(this) === arguments[1]
	}).call(thisObject, anotherFunctionNamedA, secondArgument)
})"
`;

exports[`prefer-array-index-of > invalid > function fn() {
\\u0009foo.findIndex(x => x === arguments.length)
} 1`] = `
"function fn() {
	foo.indexOf(arguments.length)
}"
`;

exports[`prefer-array-index-of > invalid > function fn() {
\\u0009foo.findIndex(x => x === this[1])
} 1`] = `
"function fn() {
	foo.indexOf(this[1])
}"
`;

exports[`prefer-array-index-of > invalid > function fn() {
\\u0009foo.findLastIndex(x => x === arguments.length)
} 1`] = `
"function fn() {
	foo.lastIndexOf(arguments.length)
}"
`;

exports[`prefer-array-index-of > invalid > function fn() {
\\u0009foo.findLastIndex(x => x === this[1])
} 1`] = `
"function fn() {
	foo.lastIndexOf(this[1])
}"
`;

exports[`prefer-array-index-of > invalid > function foo() {
\\u0009return (bar as string).findIndex(x => x === "foo");
} 1`] = `
"function foo() {
	return (bar as string).indexOf(\\"foo\\");
}"
`;

exports[`prefer-array-index-of > invalid > function foo() {
\\u0009return (bar as string).findLastIndex(x => x === "foo");
} 1`] = `
"function foo() {
	return (bar as string).lastIndexOf(\\"foo\\");
}"
`;

exports[`prefer-array-index-of > invalid > values.findIndex((x => x === (0, "foo"))) 1`] = `"values.indexOf((0, \\"foo\\"))"`;

exports[`prefer-array-index-of > invalid > values.findIndex(function (x) {return x === "foo";}) 1`] = `"values.indexOf(\\"foo\\")"`;

exports[`prefer-array-index-of > invalid > values.findIndex(x => "foo" === x) 1`] = `"values.indexOf(\\"foo\\")"`;

exports[`prefer-array-index-of > invalid > values.findIndex(x => {return x === "foo";}) 1`] = `"values.indexOf(\\"foo\\")"`;

exports[`prefer-array-index-of > invalid > values.findIndex(x => x === "foo") 1`] = `"values.indexOf(\\"foo\\")"`;

exports[`prefer-array-index-of > invalid > values.findIndex(x => x === (0, "foo")) 1`] = `"values.indexOf((0, \\"foo\\"))"`;

exports[`prefer-array-index-of > invalid > values.findIndex(x => x === foo()) 1`] = `"values.findIndex(x => x === foo())"`;

exports[`prefer-array-index-of > invalid > values.findLastIndex((x => x === (0, "foo"))) 1`] = `"values.lastIndexOf((0, \\"foo\\"))"`;

exports[`prefer-array-index-of > invalid > values.findLastIndex(function (x) {return x === "foo";}) 1`] = `"values.lastIndexOf(\\"foo\\")"`;

exports[`prefer-array-index-of > invalid > values.findLastIndex(x => "foo" === x) 1`] = `"values.lastIndexOf(\\"foo\\")"`;

exports[`prefer-array-index-of > invalid > values.findLastIndex(x => {return x === "foo";}) 1`] = `"values.lastIndexOf(\\"foo\\")"`;

exports[`prefer-array-index-of > invalid > values.findLastIndex(x => x === "foo") 1`] = `"values.lastIndexOf(\\"foo\\")"`;

exports[`prefer-array-index-of > invalid > values.findLastIndex(x => x === (0, "foo")) 1`] = `"values.lastIndexOf((0, \\"foo\\"))"`;

exports[`prefer-array-index-of > invalid > values.findLastIndex(x => x === foo()) 1`] = `"values.findLastIndex(x => x === foo())"`;
